feat: implement RFC 8628

nsklikas opened the pull request:
Implements the Device Authorization Grant to enable authentication for headless machines (see https://datatracker.ietf.org/doc/html/rfc8628) 

## Related issue(s)

Implements [RFC 8628](https://datatracker.ietf.org/doc/html/rfc8628). 

This PR is based on the work done on https://github.com/ory/hydra/pull/3252, by @supercairos and @BuzzBumbleBee. That PR was based on an older version of Hydra and was missing some features/tests.

We have prepared a [spec](https://github.com/canonical/hydra/wiki/Device-Flow-Specification), that describes our design and implementation. We have tried to mimic the existing logic in Hydra and not make changes that would disrupt the existing workflows


## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](https://github.com/ory/docs).

## Further Comments

Notes: 
- The current implementation has been manually tested only for `memory` and `postgres` databases. The tests pass all of them.
- Fosite is installed from our fork to ease testing. Once the [relevant PR](https://github.com/ory/fosite/pull/826) in fosite is merged, we will update `go.mod`.

### Testing
To test this you need to built the hydra image:
```console
make docker
```
This will create an image with the name: `oryd/hydra:latest-sqlite`

To run the flow you can use our UI, from https://github.com/canonical/identity-platform-login-ui/tree/hydra-device-test:
```console
git clone git@github.com:canonical/identity-platform-login-ui.git -b hydra-device-test
cd identity-platform-login-ui/
# The image name is hard-coded in the docker-compose file
docker compose up --remove-orphans --force-recreate -d
```

Create a client for Hydra:
```console
docker exec -it identity-platform-login-ui-hydra-1 hydra create client   --endpoint http://localhost:4445   --grant-type authorization_code,refresh_token,urn:ietf:params:oauth:grant-type:device_code --scope openid,offline_access,email,profile --token-endpoint-auth-method client_secret_post
```

Use that client to perform the device flow:
```console
docker exec -it identity-platform-login-ui-hydra-1 hydra perform device-code --client-id <client-id> --client-secret <client-secret> -e http://localhost:4444 --scope openid,offline_access,email,profile
```

The user for logging in is:
- username: `test@example.com`
- password: `test`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ory/hydra?pullRequest=3851) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ory/hydra?pullRequest=3851) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: nsklikas<br/>:x: wood-push-melon<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ory/hydra?pullRequest=3851) it.</sub>

supercairos commented:
You kept the `user_code` & `device_code` in separate tables ? 
I thought It could be merge with the `flow` table but might be tricky to do (lots of SQL constrains to manage)

zepatrik commented:
One thing I am struggling to understand is the `device_challenge` parameter. On one side, I cannot find any mention of that in the RFC, on the other side I don't see how it would be available in all variants of the device flow. Sure, when using the `verification_uri_complete` from the device authorization response, it is easy to set additional query parameters. However, the device auth flow also has to work when only the user code is entered into a generic website, like on https://youtube.com/tv/activate

zepatrik commented:
OK, I think I now cleared up the confusion I had about the `device_challenge`.
IMO it is not really necessary because it would also work to send the user straight to the UI implementation, and let the `/admin/oauth2/auth/requests/device/accept` create the flow. The relevant information at this point is always in the user-code, and only the UI provides that. However, I see how it fits better into the existing code and architecture, so I do like the proposal 👍

nsklikas commented:
Apparently the tests are broken because of https://github.com/ory/fosite/pull/827/files#diff-b92270a81f4021a9cdf52dfcfaeac9b66254471b85fd5ef4101acdbad02e4296R161, not sure if it's a bug or if the hydra tests need to be updated. @zepatrik any tip on how to fix this?

zepatrik commented:
@nsklikas I have fixed the upstream fosite issue.

I also thought about the overall flow a bit more, and I think this flow is a bit better wrt database strain (writes and storage). Also it is less complex from the bigger picture, but I agree that it might be more complex to implement. Did you consider something similar, and what are your thoughts? I would especially prefer the user and device codes to be in one table, so we can rely on the database to ensure the link between the two.
The big difference is that we would create the flow already when the device initializes everything, and then reuse it for the user as soon as we have the user code.

Also happy to discus synchronously.

[![](https://mermaid.ink/img/pako:eNp9VNuOmzAQ_ZWRH_rS7A-gKlIbtBukqqoS0SekyrEnMFqwqW0Spav9944NCbmsyoOB8Znjc2YG3oSyGkUmPP4Z0CjMSdZOdpUBvnrpAinqpQmQg_SQ44EUPm6WcbP06OCbs0e-P0LWEbI-aSc_SC8qM0alCnSQASEf38c1h6fl8jOsMygMBZIt_UXQSQtE_bBv7XGEriM0Il_QoItMn6BH58mHhAJpdMrxFzwnPEGewS90tCclA1kD5aZYTHaZIVlbzVnj-sMyvT3wVrmIBC8YYIjIYKG2vH7ZueXhjjXGogbfsJiEjmru3JfnepQX59vABRvdkumHcOs5eSgz2KAmhypECWUBRwoNrLabZ-gdi1VRxA1xWYyyVzNvL-tJDhNciK-iqcIcSkVJOR6si76UNXtynb_3deVi6HYdhbmgs8S7_j0TFyl1bHea-eIpZa9jjdKeD_GRzX63NZl5OvR_K5PAbC-ypfPbGPg9tioO70N_MxZrAplhEpwYzq3kLY88xjzgxrpOtmcCjY8dba3tQQ-OTD3akqkrnisJoUHYXX9At8P_07Yty39FA2h0b4nPNDPwapS_KoXe8-BucO_QN2PWNLyc-4FA_uDEQnTI-knzD-EtIirBkjqsRMaPWrrXSlTmnXFyCHZ7MkpkwQ24EM4OdSOyvWw9vw2pQ9OvZIq-_wPzwWta?type=png)](https://mermaid.live/edit#pako:eNp9VNuOmzAQ_ZWRH_rS7A-gKlIbtBukqqoS0SekyrEnMFqwqW0Spav9944NCbmsyoOB8Znjc2YG3oSyGkUmPP4Z0CjMSdZOdpUBvnrpAinqpQmQg_SQ44EUPm6WcbP06OCbs0e-P0LWEbI-aSc_SC8qM0alCnSQASEf38c1h6fl8jOsMygMBZIt_UXQSQtE_bBv7XGEriM0Il_QoItMn6BH58mHhAJpdMrxFzwnPEGewS90tCclA1kD5aZYTHaZIVlbzVnj-sMyvT3wVrmIBC8YYIjIYKG2vH7ZueXhjjXGogbfsJiEjmru3JfnepQX59vABRvdkumHcOs5eSgz2KAmhypECWUBRwoNrLabZ-gdi1VRxA1xWYyyVzNvL-tJDhNciK-iqcIcSkVJOR6si76UNXtynb_3deVi6HYdhbmgs8S7_j0TFyl1bHea-eIpZa9jjdKeD_GRzX63NZl5OvR_K5PAbC-ypfPbGPg9tioO70N_MxZrAplhEpwYzq3kLY88xjzgxrpOtmcCjY8dba3tQQ-OTD3akqkrnisJoUHYXX9At8P_07Yty39FA2h0b4nPNDPwapS_KoXe8-BucO_QN2PWNLyc-4FA_uDEQnTI-knzD-EtIirBkjqsRMaPWrrXSlTmnXFyCHZ7MkpkwQ24EM4OdSOyvWw9vw2pQ9OvZIq-_wPzwWta)

nsklikas commented:
We considered merging the 2 tables (I think it was also discussed in the older PR). Merging the 2 tables would complicate the schema (you would need to have 2 expiration periods, 2 `active` fields, more indexes, etc) and we would have to create new logic to handle calls to this table (now we re-use the logic that is used for all the other tokens), of course that is not a blocker but would require a decently sized refactor of this PR. We would also have to merge the 2 fosite APIs (`DeviceCodeStorage` and `UserCodeStorage`), again not a big blocker. AFAICT by merging the 2 tables we would be making 1 less read to the database (we wouldn't need to fetch the device_code in `performOAuth2DeviceVerificationFlow`) and one less write (we could invalidate the `user_code` and mark the `device_code` as ready to be used in a single query). The main drawback with the current approach is that the 2 tables (user_code and device_code) are not directly merged, instead we use the requestID to connect them. The reason we decided not to go that route is that we thought that the performance benefit does not out-weight having a uniform experience with the existing fosite APIs and hydra database, but I can see the value of changing this. 

About persisting the flow table to the database at the beginning of the flow, we would be doing one less redirect (we could send the user directly to the login UI, they wouldn't need to go through Hydra). But we would have to:
1) persist the flow to the database every time a device starts a flow (1 write)
2) fetch the flow every time a user_code is accepted (we wouldn't update it on the db, because we want to allow the user to to restart the flow)
3) update the database constraints to handle the device flow status
This was something we considered, but decided that we wouldn't be gaining much for these extra call/changes. What is the reason you think we should change this?

I would rather we keep the design as it is, because we think that these changes wouldn't improve the current flow much (I understand that, depending on the load, merging the 2 tables can offer considerable improved performance) and it would complicate the implementation. Ideally I would rather we do not change the design of the current PR (unless you think that there is something wrong with it), to get something going and to avoid getting lost on the many changes that it introduces. We can always iterate on it on subsequent PRs, BUT I understand that if we want to change the database schema (by merging the tables), it would be best to do it as early as possible to avoid having to create a migration plan.

zepatrik commented:
Thanks for revisiting, I was just adding some follow-up clarifications.

We had even more discussions also with @alnr and basically came to these conclusions:

1. We need to write the device & user codes into a table when creating them, mainly to avoid collisions. Ideally this would be one table (further `device_auth_codes`) that is only used for this purpose. The table should have the PK `(nid, device_code_sig)` and a secondary unique index `(nid, user_code_sig)`. This table will be polled by the device using the device code.
1. The flow should not be persisted in the beginning by the device, but used only by the user browser and persisted after successful completion.
2. Once the user code is used, we should mark it as such and release it by setting `user_code_sig=null`. We can then include the device code signature in the flow. The main reason here is that we can make sure the code is only used once.
3. There seems to be no value in adding the `device_challenge`. We propose to remove that. The existing CSRF token should be sufficient to ensure that a flow was completed in the same browser it was started.
4. The "accept user code API" should be an admin API (as it is now), for flexibility and consistency reasons.

Overall, the refactor to use only one table should be worth it right away. We can also help out with the refactor if necessary.

nsklikas commented:
Thanks for the quick reply.

Just to be clear, afaict refactoring the database affects both fosite and hydra. In fosite we should merge the `DeviceCodeStorage` and `UserCodeStorage` APIs into one. In hydra we should first agree on a table schema and then do the changes needed.

Regarding the unique index, I am not sure if this is the right approach. We can have a PK with `(nid, device_code_sig)`. But on the user_code index we can't have an index that involves the `nid`. The reason for this is that when the user accepts the user_code we have no other information about the flow, we don't know who initiated or what parameters were used (in my mind the `user_code` must have a one-to-one mapping to original request). If we want to support having an index like the one proposed, we would need to have different user_code verification URL per nid. E.g. for `nid=42` the verification url would be `https://hydra.example.com/oauth2/device/verify/42` and for `nid=17` the verification url would be `https://hydra.example.com/oauth2/device/verify/17`. I have to say that I don't really like this, especially taking into account that the user may have to manually write the URL down on their browser.

IMO the migration should look something like the following:
```
CREATE TABLE IF NOT EXISTS hydra_oauth2_device_auth_codes (
    device_code_signature          VARCHAR(255) NOT NULL PRIMARY KEY,
    user_code_signature          VARCHAR(255) NOT NULL,
    request_id         VARCHAR(40)  NOT NULL,
    requested_at       TIMESTAMP    NOT NULL DEFAULT NOW(),
    client_id          VARCHAR(255) NOT NULL,
    scope              TEXT         NOT NULL,
    granted_scope      TEXT         NOT NULL,
    form_data          TEXT         NOT NULL,
    session_data       TEXT         NOT NULL,
    subject            VARCHAR(255) NOT NULL DEFAULT '',
    device_code_active             BOOL         NOT NULL DEFAULT true,
    user_code_state             SMALLINT         NOT NULL DEFAULT 0,
    browser_flow_completed             BOOL         NOT NULL DEFAULT false,
    requested_audience TEXT         NULL DEFAULT '',
    granted_audience   TEXT         NULL DEFAULT '',
    challenge_id       VARCHAR(40)  NULL,
    device_code_expires_at         TIMESTAMP    NULL,
    user_code_expires_at         TIMESTAMP    NULL,
    nid                UUID         NULL,

    FOREIGN KEY (client_id, nid) REFERENCES hydra_client(id, nid) ON DELETE CASCADE,
    FOREIGN KEY (nid) REFERENCES networks(id) ON UPDATE RESTRICT ON DELETE CASCADE
    PRIMARY KEY (device_code_signature, nid)
);

CREATE INDEX hydra_oauth2_device_auth_codes_request_id_idx ON hydra_oauth2_device_auth_codes (request_id, nid);
CREATE INDEX hydra_oauth2_device_auth_codes_client_id_idx ON hydra_oauth2_device_auth_codes (client_id, nid);
CREATE INDEX hydra_oauth2_device_auth_codes_challenge_id_idx ON hydra_oauth2_device_auth_codes (challenge_id);
CREATE INDEX hydra_oauth2_device_auth_codes_challenge_id_idx ON hydra_oauth2_device_auth_codes (user_code_signature);
```

The reason there is a `user_code_state` field is that the user_code has 3 states, that is `active`(0), `used`(1) and `revoked`(2). Alternatively we could create 2 columns `user_code_active` and `browser_flow_completed`, but I think that having a single column is better.

To optimize the `user_code_signature` index, we could create a unique partial index (this should work on postgres and cockroach, we probably can find a workaround for mysql):
```
CREATE UNIQUE INDEX hydra_oauth2_device_auth_codes_challenge_id_idx ON hydra_oauth2_device_auth_codes (user_code_signature) WHERE user_code_state = 0;
```
This should be a big performance improvement on the current db queries.

@zepatrik I suggest that as a way forward:
1) If you agree with what I described, I start making the changes in fosite (shouldn't be much work). The hydra PR is blocked by the changes on fosite, so IMO this should be top priority.
2) We try to come up with a consensus on what the table schema should be, if this proves hard to do over comments/chat we can set up a call
3) I start making the changes on Hydra, unless something unexpected comes up I think it should be straight forward and I should be able to handle it on my own. If I face something blocking or it turns out that I need to refactor more parts of the code, I will ask for your help/contribution.

zepatrik commented:
I agree with the dependency here. I think now would be a good time to set up a call and discuss the schema in detail. I'll reach out via email.

zepatrik commented:
We agreed on the proposed table schema. This is the refined and detailed version of the flowchart:

[![](https://mermaid.ink/img/pako:eNp1VdtuGjEQ_ZWRH6pWJRHXEFZVpBbUgBRFFdH2oUKKzO6wa7HYW9tLSqL8e8c2l-USHpbFnjlzfM54eGOJSpFFzODfCmWCI8EzzVczCfQpubYiESWXFkbADYxwLRI834zdZmxQww-tXuj7PGTsQsabVPML6ZOZDKs8sWLNLcIo_A7PEVzd3X2FcQQTKazghXhFSD2XZ17Z_Fk79saG6LGLdsH3KFE7sE9QojbC2Ai-zfXdFbhDm_C6TcUUTKLK81VepcIJs8cm8CsYRfAbtViIhFuhJMTTSWOrDlXzSgxdjfopHhVRUWvaihsO4B4tVC7SKsgUPV3p9QmqW-OSyOXqJUQ77idixTv5NCYWdDbnn1vNZgN2jy81AlpkuQW1INEj-FkQqoO_IKavrBGksrAhqoWQS0wDUrw35MmSj54TCFlWFhYEeeLDUGNwgXR0gUchXs44gimmwtMnNeIJvAib71glOS8KlBkeFXf8nYTDQ-2S72MmPoiKX8bYFyYQRwdSbvkhdUeqBugPQ0veW1_OgNJOpETJhdArc2pPjedTNV8Je-iLI5JOou9JgqX9mGwQ8kGpJVTlJbdgvjmtjzI9yY_L1BnhDxwEplPDQqvVRw1QzzDWvZI9DyoT8nAT03NJQ1oa8ujeFy7jOfT2bjjUdT4yf_oQKNmcuhVNqaTByJHxdffNcYp5dsmo8ZS0QlYYEnz67j4NHSqNHiInlV7xYgewpfUhKx6cSrb5-8kTslM8v5SFUmRZpYXMgkUuhNKpi_wZ5_WReTzufqmioOpLlM7NUgmqKA-BtWnkGsgYumNTXJBoecgytU44I0gjljXYCun0IqW_gDcXMWNEaYUzFtFryvVyxmbyneKoM9TTRiYssrrCBtOqynIWLXhh6FflDd_-eexXabiz6I39Y1G727vuDlq3nZtet3_ba_VuGmxDy63r5qDbo53-oNnvt3rvDfaqFCE0rwe3zXa722n3ex1KaHY83B-_6Ri8_wfzHiXl?type=png)](https://mermaid.live/edit#pako:eNp1VdtuGjEQ_ZWRH6pWJRHXEFZVpBbUgBRFFdH2oUKKzO6wa7HYW9tLSqL8e8c2l-USHpbFnjlzfM54eGOJSpFFzODfCmWCI8EzzVczCfQpubYiESWXFkbADYxwLRI834zdZmxQww-tXuj7PGTsQsabVPML6ZOZDKs8sWLNLcIo_A7PEVzd3X2FcQQTKazghXhFSD2XZ17Z_Fk79saG6LGLdsH3KFE7sE9QojbC2Ai-zfXdFbhDm_C6TcUUTKLK81VepcIJs8cm8CsYRfAbtViIhFuhJMTTSWOrDlXzSgxdjfopHhVRUWvaihsO4B4tVC7SKsgUPV3p9QmqW-OSyOXqJUQ77idixTv5NCYWdDbnn1vNZgN2jy81AlpkuQW1INEj-FkQqoO_IKavrBGksrAhqoWQS0wDUrw35MmSj54TCFlWFhYEeeLDUGNwgXR0gUchXs44gimmwtMnNeIJvAib71glOS8KlBkeFXf8nYTDQ-2S72MmPoiKX8bYFyYQRwdSbvkhdUeqBugPQ0veW1_OgNJOpETJhdArc2pPjedTNV8Je-iLI5JOou9JgqX9mGwQ8kGpJVTlJbdgvjmtjzI9yY_L1BnhDxwEplPDQqvVRw1QzzDWvZI9DyoT8nAT03NJQ1oa8ujeFy7jOfT2bjjUdT4yf_oQKNmcuhVNqaTByJHxdffNcYp5dsmo8ZS0QlYYEnz67j4NHSqNHiInlV7xYgewpfUhKx6cSrb5-8kTslM8v5SFUmRZpYXMgkUuhNKpi_wZ5_WReTzufqmioOpLlM7NUgmqKA-BtWnkGsgYumNTXJBoecgytU44I0gjljXYCun0IqW_gDcXMWNEaYUzFtFryvVyxmbyneKoM9TTRiYssrrCBtOqynIWLXhh6FflDd_-eexXabiz6I39Y1G727vuDlq3nZtet3_ba_VuGmxDy63r5qDbo53-oNnvt3rvDfaqFCE0rwe3zXa722n3ex1KaHY83B-_6Ri8_wfzHiXl)

zepatrik commented on oauth2/handler.go:
```diff
@@ -689,6 +712,169 @@ func (h *Handler) getOidcUserInfo(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	// TODO evaluate if an OpenID Connect session is necessary for device flow.
```
I checked several documentation pages of other implementations, e.g. https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-device-code, and they do support the `openid` scope and ID token on the device flow.

supercairos commented:
> We agreed on the proposed table schema. This is the refined and detailed version of the flowchart:
> 
> [![](https://camo.githubusercontent.com/52a754e16074b14d0b6c9246646d47340d1578236f6bc80a17b0ba7b4ba7e0f6/68747470733a2f2f6d65726d6169642e696e6b2f696d672f70616b6f3a654e703156647475476a45515f5a5752483670574a52485845465a567042625567425246466448326f554b4b7a4f3677613748595739744c53714c38653863326c2d5553487062466e6a6c7a664d353465474f4a537046467a4f4466436d57434938457a7a56637a436651707562596945535758466b6241445978774c52493833347a645a6d785177772d7458756a37504754735173616256504d4c365a4f5a444b733873574c4e4c63496f5f41375045567a6433583246635151544b617a6768586846534432585a31375a5f466b3739736147364c474c647348334b464537734539516f6a62433241692d7a665864466268446d5f433654635555544b4c4b383156657063494a7338636d384373595266416274566949684675684a4d545453574f72446c587a536778646a666f7048685652555776616968734f34423474564337534b7367555056337039516d71572d4f53794f58714a55513737696469785476354e4359576444626e6e31764e5a674e326a793831416c706b75515731494e456a2d466b51716f4f5f494b61767242476b73724168716f57515330774455727733354d6d536a35345443466c574668594565654c4455474e7767585230675563685873343467696d6d77744d6e4e65494a764169623731676c4f53384b6c426b65465866386e595444512d325337324d6d506f694b58386259467959515277645362766b6864556571427567505130766557315f4f674e4a4f7045544a68644172633270506a6564544e56384a652d694c49354a4f6f75394a677158396d477751386b47704a56546c4a626467766a6d746a7a493979595f4c31426e6844787745706c504451717656527731517a7a4457765a493944796f54386e415430334e4a51316f6138756a654679376a4f665432626a6a5564543479665f6f514b4e6d637568564e71615442794a48786466664e6359703564736d6f385a5330516c5959456e7a36376a344e4853714e4869496e6c56377859676577706655684b783663537262352d386b54736c4d3876355346556d525a7059584d676b5575684e4b70695f775a355f575265547a7566716d696f4f704c6c4d374e55676d714b412d4274576e6b47736759756d4e54584a426f65636779745534344930676a6c6a585943756e304971575f674463584d574e456159557a4674467279765679786d62796e654b6f4d395454526959737372724342744f71796e49574c5868683646666c44645f2d656578586162697a364933395931473732377675446c71336e5a746574335f62615f5675476d784479363372357144626f35332d6f4e6e76743372764466617146434530727765337a58613732326e336578314b6148593833422d5f365269385f77667a4869586c3f747970653d706e67)](https://mermaid.live/edit#pako:eNp1VdtuGjEQ_ZWRH6pWJRHXEFZVpBbUgBRFFdH2oUKKzO6wa7HYW9tLSqL8e8c2l-USHpbFnjlzfM54eGOJSpFFzODfCmWCI8EzzVczCfQpubYiESWXFkbADYxwLRI834zdZmxQww-tXuj7PGTsQsabVPML6ZOZDKs8sWLNLcIo_A7PEVzd3X2FcQQTKazghXhFSD2XZ17Z_Fk79saG6LGLdsH3KFE7sE9QojbC2Ai-zfXdFbhDm_C6TcUUTKLK81VepcIJs8cm8CsYRfAbtViIhFuhJMTTSWOrDlXzSgxdjfopHhVRUWvaihsO4B4tVC7SKsgUPV3p9QmqW-OSyOXqJUQ77idixTv5NCYWdDbnn1vNZgN2jy81AlpkuQW1INEj-FkQqoO_IKavrBGksrAhqoWQS0wDUrw35MmSj54TCFlWFhYEeeLDUGNwgXR0gUchXs44gimmwtMnNeIJvAib71glOS8KlBkeFXf8nYTDQ-2S72MmPoiKX8bYFyYQRwdSbvkhdUeqBugPQ0veW1_OgNJOpETJhdArc2pPjedTNV8Je-iLI5JOou9JgqX9mGwQ8kGpJVTlJbdgvjmtjzI9yY_L1BnhDxwEplPDQqvVRw1QzzDWvZI9DyoT8nAT03NJQ1oa8ujeFy7jOfT2bjjUdT4yf_oQKNmcuhVNqaTByJHxdffNcYp5dsmo8ZS0QlYYEnz67j4NHSqNHiInlV7xYgewpfUhKx6cSrb5-8kTslM8v5SFUmRZpYXMgkUuhNKpi_wZ5_WReTzufqmioOpLlM7NUgmqKA-BtWnkGsgYumNTXJBoecgytU44I0gjljXYCun0IqW_gDcXMWNEaYUzFtFryvVyxmbyneKoM9TTRiYssrrCBtOqynIWLXhh6FflDd_-eexXabiz6I39Y1G727vuDlq3nZtet3_ba_VuGmxDy63r5qDbo53-oNnvt3rvDfaqFCE0rwe3zXa722n3ex1KaHY83B-_6Ri8_wfzHiXl)



At the end of the User-Browser login flow part, It's good to redirect the user to a "post device login webpage" so the user knows he can safely close his Browser webpage and that the device is authenticated succesfully. 

I don't see this being mentioned, Is this handled ?

nsklikas commented on driver/config/provider.go:
```diff
@@ -392,6 +399,26 @@ func (p *DefaultProvider) fallbackURL(ctx context.Context, path string, host str
+	return p.p.DurationF(KeyDeviceAuthTokenPollingInterval, time.Second*5)
+}
+
+// GetUserCodeLength returns configured user_code length
+func (c *DefaultProvider) GetUserCodeLength(ctx context.Context) int {
+	return 8
+}
+
+// GetDeviceAuthTokenPollingInterval returns configured user_code allowed symbols
+func (c *DefaultProvider) GetUserCodeSymbols(ctx context.Context) []rune {
+	return []rune(randx.AlphaUpper)
+}
```
I would rather leave this as a follow-up. If you want we can compile a list of improvements and I can open an issue to keep track of them.

nsklikas commented on oauth2/handler.go:
```diff
@@ -689,6 +712,167 @@ func (h *Handler) getOidcUserInfo(w http.ResponseWriter, r *http.Request) {
+	//   - the granted audiences
+	//   - the user_code_state set to accepted
+	// This marks it as ready to be used for the token exchange endpoint.
+	err = h.r.OAuth2Storage().UpdateDeviceCodeSessionByRequestID(ctx, f.DeviceCodeRequestID.String(), req)
```
This is a bit awkward. When the user successfully logs in, there are 3 writes happening to the database all by this function. First the flow is written in the database (in [consent/strategy_default.go](https://github.com/ory/hydra/pull/3851/files#diff-265e72885dd37a9d8c6f81e2c928b2bfad035cecd98051ef2f04ea88a56ed540R1245)), the device auth request is updated and then the oidc session is updated.

I would like to have them grouped in a transaction. Currently there is the edge case where:
- 2 browser tabs open the device verification URL and input the same code
- Each user agent will get a different flow
- If both flows are completed, the one that got there first will see the device complete page and the other one will get an error

This will result in both flows being persisted to the database. If we were to wrap the flow creation and the device code update in a single transaction we would solve this issue, but the current code structure makes it difficult to implement (I don't want to start a transaction every time `HandleOAuth2DeviceAuthorizationRequest` is called, as most of the time nothing will be written to the database)

zepatrik replied:
I see, so it is mainly hard because you cannot tell whether the stateless flow will be updated (no DB write), or the final flow will be updated in the DB, correct?
The consequence of the flow being written is not security-relevant, right? It's rather a UX improvement? I think in that case we can fix that later.

nsklikas replied:
No I don't think it has any security implications. I will just add a TODO on the code so that this does not get lost

nsklikas commented:
I updated the database schema and refactored the logic a little. I tested it on postgres with ~1M rows and it looks like all queries are using the indexes:
![image](https://github.com/user-attachments/assets/a4815ffa-ea10-4f9c-b642-57b171c682a5)


@zepatrik please have another look when you can

aeneasr reviewed (commented):
Thank you for the PR. I've looked at the surface issues first in this review, a full review of functionality and security is outstanding.

nsklikas commented on consent/handler.go:
```diff
@@ -1047,11 +1051,150 @@ func (h *Handler) getOAuth2LogoutRequest(w http.ResponseWriter, r *http.Request,
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithHint(`'user_code' signature could not be computed`)))
+		return
+	}
+	userCodeRequest, err := h.r.OAuth2Storage().GetUserCodeSession(r.Context(), userCodeSignature, nil)
```
AFAICT the only database call happening in this handler is fetching the user_code session. The flow is not persisted and the user code is not updated. I don't think we need a transaction, or am I missing something?

nsklikas commented on consent/handler.go:
```diff
@@ -1047,11 +1051,150 @@ func (h *Handler) getOAuth2LogoutRequest(w http.ResponseWriter, r *http.Request,
+	if err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrTokenExpired.WithWrap(err).WithHint(`'user_code' has expired`)))
+		return
+	}
```
Do you mean that fosite should implement the handler that is used for accepting the `user_code` (or at least provide some helper function)? 

It's something that I wasn't sure how to approach. I decided that since the RFC does not describe how the `user_code` should be accepted, fosite should not provide any helper functions for that.

nsklikas commented on persistence/sql/persister_device.go:
```diff
@@ -0,0 +1,305 @@
+		return err
+	}
+
+	stmt := fmt.Sprintf(
```
Not sure I understand, afaict all update statements are generated this way. Or do you mean to exec the query in an one-liner to avoid the creation of the `stmt` var?

nsklikas commented on consent/handler.go:
```diff
@@ -1047,11 +1051,150 @@ func (h *Handler) getOAuth2LogoutRequest(w http.ResponseWriter, r *http.Request,
+
+	challenge := stringsx.Coalesce(
+		r.URL.Query().Get("device_challenge"),
+		r.URL.Query().Get("challenge"),
```
Fair, I will remove `challenge`

nsklikas commented on consent/handler.go:
```diff
@@ -1047,11 +1051,150 @@ func (h *Handler) getOAuth2LogoutRequest(w http.ResponseWriter, r *http.Request,
+	f.Client, err = h.r.ClientManager().GetConcreteClient(ctx, f.ClientID)
+	if err != nil {
+		return nil, err
+	}
```
Good catch, thank you

nsklikas commented:
After some out-of-band discussions, I have created a new PR (https://github.com/ory/hydra/pull/3912) using the same commits but from a different fork. Please let me know if you think that the comments should be moved to the new PR or not.
